SPI Engine: create inverted CS mode (#1301)
analogdevicesinc/hdl: docs/library/spi_engine/instruction-format.rst

================================================================================

The SPI Engine instruction set is a simple 16-bit instruction set of which

> 13-bit is currently allocated (bits 15,11,10 are always 0).

Reviewer: ```suggestion
13 bits are currently allocated (bits 15,11,10 are always 0).
```

better grammar

The chip-select instruction updates the value chip-select output signal of the
SPI Engine execution module.

> The actual values on each pin might be inverted relative to the command

Reviewer: ```suggestion
The physical outputs on each CS pin may be inverted relative to the CS command
```

according to the mask set by :ref:`spi_engine cs-invert-mask-instruction`. The
Invert Mask acts only on the output registers of the Chip-Select pins. Thus, if

> the last 8 bits of this instruction are 0xFE, CS[0] will be active regardless of

Reviewer: Wouldn't 0xFE invert all outputs except CS[0]?

Author: This text is for the Chip-Select instruction, not the CS Invert Mask Instruction. The idea is that you use the Chip Select instruction the same way, for any Invert Mask state. The values are only inverted at the output. So 0xFE will always activate CS[0]. If the Invert Mask is the default 0x00, then only CS[0] will be low. If the mask is 0xFF, then only CS[0] will be high. If the mask is 0x01, then all CS will be high (but only CS[0] is active, since it's the only one that's inverted)

I think I'll add these examples to the description.

Author: This text is for the Chip-Select instruction, not the CS Invert Mask Instruction. The idea is that you use the Chip Select instruction the same way, for any Invert Mask state. The values are only inverted at the output. So 0xFE will always activate CS[0]. If the Invert Mask is the default 0x00, then only CS[0] will be low. If the mask is 0xFF, then only CS[0] will be high. If the mask is 0x01, then all CS will be high (but only CS[0] is active, since it's the only one that's inverted)

I think I'll add these examples to the description.

Reviewer: Ah, ok. It just says "this instruction" and the last instruction mentioned is the invert instruction, so it sounds like "this" is referring to the invert instruction. So this could be made more clear that is is referring to a different instruction.

Author: You are absolutely right, could be more clear. I tried to make it less ambiguos on the commit below

> the last 8 bits of this instruction are 0xFE, CS[0] will be active regardless of
> polarity, and transparently to the programmer.

Reviewer: ```suggestion
the last 8 bits of this instruction are 0x01, CS[0] will be active high instead of active low.
```

The chip-select instruction updates the value chip-select output signal of the
SPI Engine execution module.

The physical outputs on each pin may be inverted relative to the command
according to the mask set by :ref:`spi_engine cs-invert-mask-instruction`. The
Invert Mask acts only on the output registers of the Chip-Select pins. Thus, if
the last 8 bits of the Chip-Select instruction are 0xFE, only CS[0] will be

> active regardless of polarity, and transparently to the programmer.

Reviewer 2: `and transparent to the programmer` or `and transparent to the user`

     - Time
     - The amount of time to wait.

.. _spi_engine cs-invert-mask-instruction:

CS Invert Mask Instruction
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

== == == == == == = = = = = = = = = =
15 14 13 12 11 10 9 8 7 6 5 4 3 2 1 0
== == == == == == = = = = = = = = = =
0  1  0  0  r  r  r r m m m m m m m m
== == == == == == = = = = = = = = = =

The CS Invert Mask Instructions allows the user to select on a per-pin basis
whether the Chip Select will be active-low (default) or active-high (inverted).
Note that the Chip-Select instructions should remain the same because the value

> of CS is inverted at the output register, and additional logic (eg. reset

Reviewer 2: `e.g.` or `e.g.:`

counters) occurs when the CS active state is asserted. 

Since the physical values on the pins are inverted at the output, the current

> Invert Mask should not affect the use of the :ref:`spi_engine cs-instruction`.

Reviewer 2: ```suggestion
Invert Mask does not affect the :ref:`spi_engine cs-instruction`.
```

   * - Bits
     - Name
     - Description
   * - [7:4]
     - reserved
     - Must always be 0.
   * - [3]
     - cs_active_high
     - Configures the polarity of CS. When 0 (default), CS is active-low as 
     - usual. If set to 1, CS will be active-high. Note that there's no need 
     - to change the commands: for cs_active_high=1, the value of CS will be 

>      - automatically inverted with respect to the command.

Reviewer: It is helpful to driver writers to include the version a feature was added. So suggest adding here:

This feature was added in vX.Y.Z. Setting this bit in older versions has no effect.

   * - Bits
     - Name
     - Description
   * - [7:4]
     - reserved
     - Must always be 0.
   * - [3]
     - cs_active_high

>      - Configures the polarity of CS. When 0 (default), CS is active-low as 
>      - usual. If set to 1, CS will be active-high. Note that there's no need 
>      - to change the commands: for cs_active_high=1, the value of CS will be 
>      - automatically inverted with respect to the command. This was introduced
>      - in version 1.02.71 of the core, setting this bit on earlier versions has
>      - no effect.

Reviewer 2: Long paragraphs are not prefixed by `-`
```suggestion
     - Configures the polarity of CS. When 0 (default), CS is active-low as 
       usual. If set to 1, CS is active-high. Note that the command instructions
       remain unchanged because the value of CS is inverted at the output
       register and additional logic occurs when the CS active state is asserted
       (e.g., reset counters).
       This was introduced in version 1.02.71 of the core, setting this bit on
       earlier versions has no effect.
```
Rephrased to unhide reset states at `CMD_CHIPSELECT` and use imperative grammar.
